tremor-connectors-azure/src/monitor/ingest/writer.rs

```diff
@@ -0,0 +1,428 @@
+use tremor_connectors::Connector;
+use tremor_connectors::ConnectorBuilder;
+use tremor_connectors::ConnectorType;
+use tremor_system::killswitch::KillSwitch;
+use tremor_system::qsize;
+
+mod sink;
+mod source;
+
+// The Azure Monitor Ingestion API requires the following scope for default permissions
+// to be available for the DCR based ingestion endpoints to be used.
+//
+const AZURE_MONITOR_AUTH_SCOPE: &str = "https://monitor.azure.com/.default";
+
+#[derive(Deserialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub(crate) struct Config {
+    /// Azure authentication configuration. Required. The provided scope determines
+    /// the permissions available to the token.
+    pub auth: AuthConfig,
+
+    /// The Azure Data Collection Rule Base URL. Required.
+    pub dce_base_url: String,
+
+    /// The immutable Data Collection Rule ID. Required.
+    pub dcr: String,
+
+    /// The stream name to publish to, from the Data Collection Rule. Required.
+    pub stream: String,
+
+    /// Api Version. Default value is "2023-01-01".
+    #[serde(default = "default_api_version")]
+    pub api_version: String,
+
+    /// Concurrency capacity limits ( in flight requests )
+    #[serde(default = "default_concurrency")]
+    pub(super) concurrency: usize,
+
+    /// Timeout for the request in milliseconds. Optional.
+    timeout: Option<u64>,
+}
+
+const DEFAULT_CONCURRENCY: usize = 4;
+
+fn default_concurrency() -> usize {
+    DEFAULT_CONCURRENCY
+}
+
+fn default_api_version() -> String {
+    "2023-01-01".to_string()
+}
+
+impl tremor_config::Impl for Config {}
+
+struct RequestBuilder {
+    id: RequestId,
+    rest_endpoint: azure_core::Url,
+    auth_token: Option<String>,
+    data: Vec<Vec<u8>>,
+}
+
+impl RequestBuilder {
+    async fn new(
+        id: RequestId,
+        _meta: Option<&Value<'_>>,
+        config: &Config,
+    ) -> anyhow::Result<Self> {
+        let rest_endpoint = format!(
+            "{dce_base_url}/dataCollectionRules/{dcr}/streams/{stream}?api-version={api_version}",
+            dce_base_url = config.dce_base_url,
+            dcr = config.dcr,
+            stream = config.stream,
+            api_version = config.api_version
+        );
+        let azure_core_url = azure_core::Url::parse(&rest_endpoint)?;
+
+        Ok(Self {
+            id,
+            rest_endpoint: azure_core_url,
+            auth_token: Some(config.auth.get_token().await?),
+            data: vec![],
+        })
+    }
+
+    pub(super) async fn append<'event>(
+        &mut self,
+        value: &'event Value<'event>,
+        meta: &'event Value<'event>,
+        ingest_ns: u64,
+        serializer: &mut EventSerializer,
+    ) -> anyhow::Result<()> {
+        let chunks = serializer
+            .serialize_for_stream_with_codec(value, meta, ingest_ns, self.id.get(), None)
+            .await?;
+        self.append_data(chunks);
+        Ok(())
+    }
+
+    fn append_data(&mut self, chunks: Vec<Vec<u8>>) {
```

Comment:
This is probably fine for a experimental connector but to release it we'll want a streaming sender. Let's add a TODO so we remember?

Reply:
I didn't find a way nice way to make it streaming so punted for now. I'll add a TODO as it makes sense to do some live testing for this